src/XGridCtrl.cpp

```diff
@@ -912,7 +912,7 @@ XGridCtrl::MakeVisible(const puz::Square & square)
 
 
 bool
-XGridCtrl::SetSquareText(puz::Square & square, const wxString & text)
+XGridCtrl::SetSquareText(puz::Square & square, const wxString & text, bool propagate)
```

Comment:
TL;DR -- slight preference for doing this in puz::Square, but not a super strong one.

> If you still think it's worthwhile, would you see this as a new field on Square to contain any partner square that square might have?

Yeah, that's what I'm thinking. puz::Square already has a number of pointers to other squares that are specifically used for speeding up lookups to neighboring squares (`m_next`). This seems like a similar enough concept that it could be worth storing `m_partner`, even if all it does is replace the loop in XGridCtrl with `square->GetPartner()`.

The main reason I think keeping the state in sync would be a good idea is that effectively these are views of the _same_ square, not just two squares that happen to be connected . . . although granted in this case one of them has a different set of corner marks than the other. I think it cuts down the surface area for potential bugs if they're automatically kept in sync, but I can definitely see an argument for treating them as separate squares.

> When we're propagating in the body of this function, we make a recursive call back to this method. If we don't set propagate to false there, we'd infinitely recurse back and forth. So this is a simple way to short circuit after propagating once.

True, although you could add the extra arg to puz::Square::SetText as well, or potentially just do the extra recursive call, but then make it a no-op if the old text is the same as the new text (though that might be ugly).

> If you're doing a grid-wide operation, like erase or reveal all, you could turn propagation off or call direct setters on squares to avoid a bit of extra work finding partner squares since the parent loop is going to reach them eventually.

My guess is that if the lookup is cheap enough the one extra call per square wouldn't be so bad, although it would be nice to not have to do double the work.

Reply:
Uploaded another commit to move the partner handling down to puz::Square to the extent possible. XGridCtrl still needs to know when to refresh the partner square UI when making edits, but otherwise the actual data mutations happen at the data layer. Let me know what you think!

Reply:
Yeah, that looks great. Makes sense that there might be some drawing functions that need the recursive flag.